### PR TITLE
[ML-26284] Use DateOffset in foreast utils to support calculation based on months quaters and years.

### PR DESCRIPTION
**Pull request by shitaoli-db**

### CODE REVIEW

**Review by yxiong — COMMENTED** [review]
I'll leave this review to @lu-wang-dl :)

File: runtime/databricks/automl_runtime/forecast/__init__.py [hunk 1]
```diff
@@ -33,5 +34,28 @@
...
     "sec": "S",
-    "second": "S"
+    "second": "S",
+    'M': 'MS',
+    'MS': 'MS',
+    'months': 'MS',
+    'month': 'MS',
+    'Q': 'QS',
+    'QS': 'QS',
+    'quaters': 'QS',
```
Comment (lu-wang-dl) [inline runtime/databricks/automl_runtime/forecast/__init__.py:44]:
```suggestion
    'quarters': 'QS',
```

File: runtime/databricks/automl_runtime/forecast/__init__.py [hunk 2]
```diff
@@ -33,5 +34,28 @@
...
+    'quater': 'QS',
```
Comment (lu-wang-dl) [inline runtime/databricks/automl_runtime/forecast/__init__.py:45]:
```suggestion
    'quarter': 'QS',
```

File: runtime/databricks/automl_runtime/forecast/utils.py [hunk 3]
```diff
@@ -31,17 +32,27 @@ def get_validation_horizon(df: pd.DataFrame, horizon: int, unit: str) -> int:
...
     else:
-        validation_horizon_timedelta = df_timedelta / MIN_HORIZONS
-        validation_horizon = validation_horizon_timedelta // pd.to_timedelta(1, unit=unit)
-        _logger.info(f"Horizon {horizon_timedelta} too long relative to dataframe's timedelta. Validation horizon will be reduced to {validation_horizon_timedelta}.")
-        return validation_horizon
+        # In order to calculate the validation horizon, we incremental add offset to the start time to
+        # the quater of total timedelta.
+        unit_dateoffset = pd.DateOffset(**dict({DATE_OFFSET_KEYWORD_MAP[unit] : 1}))
+        if unit == 'QS':
+            unit_dateoffset = unit_dateoffset*3
```
Comment (lu-wang-dl) [inline runtime/databricks/automl_runtime/forecast/utils.py:47]: Will it be better to calculate quarter data with something like
```
DateOffset(months=3)
```
Are these equivalent: `DateOffset(months=3)` and `DateOffset(months=1)*3`?

File: runtime/tests/automl_runtime/forecast/prophet/model_test.py [hunk 4]
```diff
@@ -66,6 +66,10 @@ def test_model_save_and_load(self):
 
     def test_make_future_dataframe(self):
         for feq_unit in OFFSET_ALIAS_MAP:
+            # Temporally disable the year month and quater since we
```
Comment (lu-wang-dl) [inline runtime/tests/automl_runtime/forecast/prophet/model_test.py:69]:
```suggestion
            # Temporally disable the year, month and quarter 
```

File: runtime/databricks/automl_runtime/forecast/utils.py (hunk 3 above)
Comment (shitaoli-db) [reply runtime/databricks/automl_runtime/forecast/utils.py:47]: Yes, these are exactly equivalent.

File: runtime/tests/automl_runtime/forecast/prophet/model_test.py (hunk 4 above)
Comment (shitaoli-db) [reply runtime/tests/automl_runtime/forecast/prophet/model_test.py:69]: Done.

**Comment by shitaoli-db** [comment]
gentle ping.

**Review by wenfeiy-db — APPROVED** [review]
LGTM! Left some minor comments.

File: runtime/databricks/automl_runtime/forecast/utils.py [hunk 5]
```diff
@@ -31,17 +32,27 @@ def get_validation_horizon(df: pd.DataFrame, horizon: int, unit: str) -> int:
...
+        max_horizon = 0
+        cur_timestamp = df["ds"].min()
+        while cur_timestamp + unit_dateoffset <= df["ds"].max():
+            cur_timestamp += unit_dateoffset
+            max_horizon += 1
+        _logger.info(f"Horizon {horizon_dateoffset} too long relative to dataframe's "
+        f"timedelta. Validation horizon will be reduced to {max_horizon//MIN_HORIZONS*unit_dateoffset}.")
+        return max_horizon // MIN_HORIZONS
```
[the remark this answers is not in the document]
Comment (shitaoli-db) [reply runtime/databricks/automl_runtime/forecast/utils.py]: Yes, dateoffset does not support divide. Will add a comment.

File: runtime/databricks/automl_runtime/forecast/__init__.py [hunk 6]
```diff
@@ -33,5 +34,28 @@
...
+    'quarter': 'QS',
+    'Y': 'YS',
+    'YS': 'YS',
+    'years': 'YS',
+    'year': 'YS',
+}
+
+DATE_OFFSET_KEYWORD_MAP = {
+    'YS': 'years',
+    'QS': 'months',
```
[the remark this answers is not in the document]
Comment (shitaoli-db) [reply runtime/databricks/automl_runtime/forecast/__init__.py:54]: Done.

File: runtime/databricks/automl_runtime/forecast/__init__.py [hunk 7]
```diff
@@ -14,6 +14,7 @@
 # limitations under the License.
 #
 
+# Offset Alias reference: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases
 OFFSET_ALIAS_MAP = {
```
[the remark this answers is not in the document]
Comment (shitaoli-db) [reply runtime/databricks/automl_runtime/forecast/__init__.py:18]: Nope, https://github.com/databricks/universe/blob/master/automl/python/databricks/automl/internal/forecast/forecast.py we import it here.

File: runtime/databricks/automl_runtime/forecast/utils.py [hunk 8]
```diff
@@ -31,17 +32,27 @@ def get_validation_horizon(df: pd.DataFrame, horizon: int, unit: str) -> int:
     :param unit: frequency unit of the time series, which must be a pandas offset alias
     :return: horizon used for validation, in terms of the input `unit`
     """
-    MIN_HORIZONS = 4 # minimum number of horizons in the dataframe
-    df_timedelta = df["ds"].max() - df["ds"].min()
-    horizon_timedelta = pd.to_timedelta(horizon, unit=unit)
+    MIN_HORIZONS = 4 # minimum number of horizons in the datafram
+    horizon_dateoffset = pd.DateOffset(**dict({DATE_OFFSET_KEYWORD_MAP[unit] : horizon}))
```
[the remark this answers is not in the document]
Comment (shitaoli-db) [reply runtime/databricks/automl_runtime/forecast/utils.py:36]: Done, thanks for suggesting!

**Pull request merged**